### PR TITLE
Feature: Clusters endpoint is able to return Custom auth metadata info

### LINKED ISSUE #21447: 🚀 Feature: pass custom auth metadata to frontend
### 🔖 Feature description

There should be an extensible way to add new fields to data returned from this endpoint:

https://github.com/backstage/backstage/blob/b5ae2e5a62d61f188e1f7b6ac4aeabb8429c6f89/plugins/kubernetes-backend/src/service/KubernetesBuilder.ts#L364-L378

So that integrators writing custom auth strategies with frontend logic can reliably communicate the necessary config data to the frontend from the backend.

### 🎤 Context

With custom auth strategies introduced in #19903 and ensured to be invoked whenever Kubernetes data is requested in #20171, there is a gap for integrators: if you wish to write custom strategies which involve frontend logic which depends on custom per-cluster authMetadata fields, there is no way to communicate that metadata from the backend to the frontend at present.

### ✌️ Possible Implementation

The solution that suggests itself to me is that the `AuthenticationStrategy` interface could have a new method like `presentAuthMetadata(authMetadata: AuthMetadata): AuthMetadata`, which would effectively filter the set of metadata fields which should be included in responses from the `/api/kubernetes/clusters` endpoint. For every cluster, the API handler could find the `AuthenticationStrategy` corresponding to its `authProvider` value and invoke this method, embedding the result in an `auth` field on the relevant JSON object in the response envelope.

### 👀 Have you spent some time to check if this feature request has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by andmagom**
## Hey, I just made a Pull Request!

This PR closes #21447 

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by jamieklassen** [comment]
FWIW I just tested this out by using

```typescript
.addAuthStrategy('myauth', {
  getCredential: async _ => ({ type: 'anonymous' }),
  validateCluster: _ => [],
  presentAuthMetadata: x => x,
})
```

and saw a `/clusters` response like this:

```json
{
  "items": [
    {
      "name": "oidc-cluster",
      "authProvider": "myauth",
      "auth": {
        "kubernetes.io/api-server": "https://127.0.0.1:53725",
        "kubernetes.io/auth-provider": "myauth",
        "kubernetes.io/skip-metrics-lookup": "true"
      }
    }
  ]
}
```

so I think this successfully closes #21447 !

File: .changeset/kind-clouds-fly.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+---
+'@backstage/plugin-kubernetes-backend': patch
+'@backstage/plugin-kubernetes-node': patch
+---
+
+Enabled a way to include custom auth metadata info on the clusters endpoint
```
[the remark this answers is not in the document]
Comment (andmagom) [reply .changeset/kind-clouds-fly.md]: Done !

File: .changeset/kind-clouds-fly.md [hunk 2]
```diff
@@ -0,0 +1,6 @@
+---
+'@backstage/plugin-kubernetes-backend': patch
+'@backstage/plugin-kubernetes-node': patch
+---
+
+Enabled a way to include custom auth metadata info on the clusters endpoint. Ff you want to implement a Kubernetes auth strategy which requires surfacing custom auth metadata to the frontend, use the new presentAuthMetadata method on the AuthenticationStrategy interface
```
Comment (jamieklassen) [inline .changeset/kind-clouds-fly.md:6]: Typo
```suggestion
Enabled a way to include custom auth metadata info on the clusters endpoint. If you want to implement a Kubernetes auth strategy which requires surfacing custom auth metadata to the frontend, use the new presentAuthMetadata method on the AuthenticationStrategy interface.
```

File: plugins/kubernetes-backend/src/service/KubernetesBuilder.test.ts [hunk 3]
```diff
@@ -205,6 +209,145 @@ describe('KubernetesBuilder', () => {
...
     });
+
+    it('happy path: lists clusters with custom AuthStrategy and custom auth metadata', async () => {
+      objectsProviderMock = {
+        getKubernetesObjectsByEntity: jest
+          .fn()
+          .mockResolvedValue(happyK8SResult),
+        getCustomResourcesByEntity: jest.fn().mockImplementation(_ => {
+          return Promise.resolve(happyK8SResult);
+        }),
```
Comment (jamieklassen) [inline plugins/kubernetes-backend/src/service/KubernetesBuilder.test.ts:220]: Maybe you want to use this pattern everywhere?
```suggestion
        getCustomResourcesByEntity: jest.fn().mockResolvedValue(happyK8SResult),
```

File: plugins/kubernetes-backend/src/service/KubernetesBuilder.test.ts [hunk 4]
```diff
@@ -205,6 +209,145 @@ describe('KubernetesBuilder', () => {
...
+      };
+
+      const clusterSupplierMock = {
+        getClusters: jest.fn().mockImplementation(_ => {
+          return Promise.resolve([
```
Comment (jamieklassen) [inline plugins/kubernetes-backend/src/service/KubernetesBuilder.test.ts:225]: can use `mockResolvedValue` here too

**Review by jamieklassen — APPROVED** [review]
Nice! Plus some little non-blocking ideas

File: plugins/kubernetes-backend/src/service/KubernetesBuilder.test.ts (hunk 4 above)
Comment (andmagom) [reply plugins/kubernetes-backend/src/service/KubernetesBuilder.test.ts]: Done !

**Comment by jamieklassen** [comment]
I'm sure I gave you some conflicts when I merged #22233, sorry! But if you rebase and incorporate #22297, the E2E tests will pass!

**Comment by jamieklassen** [comment]
Linter failing here because a test has no assertions, plus the typechecker is failing due to an unused variable in the builder tests; maybe a problem with the rebase?

**Pull request merged**